rfcs/0053-reducing-daemon-memory-usage.md

```diff
@@ -0,0 +1,137 @@
+## Summary
+[summary]: #summary
+
+This RFC proposes changes to the Berkeley release of the daemon which will bring daemon's maximum possible memory usage within the range of our current hardware memory requirements of the Mainnet release.
+
+## Motivation
+[motivation]: #motivation
+
+With zkApps enabled, the maximum memory usage of the daemon now exceeds the current hardware requirements (`16GB`), in the event that the network is fully saturated with max-cost zkApps transactions in blocks for an extended period of time. With the planned parameters for Berkeley, we estimate that the maximum memory usage of a daemon is around `58.144633GB`.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+In order to reduce the memory usage of the daemon, we will offload some of the memory allocated data to an on-disk cache. The [memory analysis section](#memory-analysis) will show that the majority of memory in the fully saturated max-cost transaction environment comes from ledger proofs, zkApps proofs, and zkApps verification keys. Each of these pieces of data are accessed infrequently by the daemon, and, as such, do not need to be stored in RAM for fast access.
+
+Below is a list of all the interactions the daemon has with ledger proofs, zkApps proofs, and zkApps verification keys:
```

Comment:
Are we going to read/write to LMDB synchronously? If not, do any of these usages require refactoring to move to an async context?

Reply:
This is actually what's nice about LMDB from the scheduler perspective: because the db is memory mapped, the process just does a sync operation on memory, and the kernel handles the I/O. For writes, this means the synchronous cost to the process is just a memory write. For reads, assuming you miss the cache, the cost is the time it takes for the kernel to load that page from disk. But the analysis above shows that this cost is negligible even for random reads, so there is no need to make these reads and writes async like you would need to when using other databases.